### PR TITLE
Cleanup and some refactoring

### LINKED ISSUE #83: Multiple Memcached instances using unix sockets
There is a need to set up several Memcached instances using unix sockets.

Currently the cookbook supports such a deployment scenario only through setting extra_cli_options ("-a" and "-s" parameters).

It could be useful to support socket configuration resource properties.

### LINKED ISSUE #102: Run latest cookstyle
### :speaking_head: Foreword

Thank for taking the time to fill this bug report fully. Without it we may not be able to fix the bug, and the issue may be closed without resolution.

## :ghost: Brief Description

Chef have released updated cookstyle rules, we should therefore run the auto fix against the cookbook

1. Ensure you are on the latest stable `chef-workstation`
1. Run `cookstyle -a`

Issue sprouted from sous-chefs/meta/issues/111. If not applicable then issue should be closed.

### LINKED ISSUE #109: Library helper functions need to be properly namespaced
### :speaking_head: Foreword

The library helper functions are global and conflict with other cookbooks such as apache2.

## :ghost: Brief Description

Some helper functions such as ``lock_dir`` conflict with the apache2 cookbook which can cause problems when both cookbooks are included in a run list. The primary issue is that the helper functions are not properly namespaced like the apache2 cookbook is.

The resolution is to properly namespace the helper functions so they don't get pulled in globally.

## :pancakes: Cookbook version

6.0.0

## :woman_cook:  Chef-Infra Version

15

## :tophat: Platform details

Ubuntu/Debian/CentOS

### PR DESCRIPTION
**Pull request by ramereth**
# Description

This fixes a variety of issues and also improves some functionality.

## Issues Resolved

- Resolves #83 
- Resolves #109 
- Resolves #102 

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big Pull Request.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -40,11 +40,13 @@ jobs:
     strategy:
       matrix:
         os:
-          - 'amazonlinux'
```
Comment (tas50) [inline .github/workflows/ci.yml:43]: you can do amazonlinux-2 here to get testing on the RHEL7 based amazon linux

File: recipes/_package.rb [hunk 2]
```diff
@@ -16,18 +17,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+class ::Chef::Recipe
```
Comment (tas50) [inline recipes/_package.rb:20]: You want to include this via the library not not via the recipe. It makes it a lot more clear that you're injecting into a global space.

From chef-client cookbook
Chef::DSL::Recipe.send(:include, Opscode::ChefClient::Helpers)
Chef::Resource.send(:include, Opscode::ChefClient::Helpers)

File: recipes/default.rb [hunk 3]
```diff
@@ -22,6 +22,7 @@
 # this recipe simply uses the memcached_instance custom resource
 # for additional customization you can use this resource in your own wrapper cookbook
 memcached_instance 'memcached' do
+  extend Memcached::Helpers
```
Comment (tas50) [inline recipes/default.rb:25]: You def don't want to do this on the resource declaration.

File: resources/memcached_instance.rb [hunk 4]
```diff
@@ -17,16 +18,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+include Memcached::Helpers
 
-provides :memcached_instance_sysv_init
-
-provides :memcached_instance, os: 'linux'
+resource_name :memcached_instance
```
Comment (tas50) [inline resources/memcached_instance.rb:23]: you probably want to also provide the legacy name so things don't break for all the users of the systemd resource

File: libraries/helpers.rb [hunk 5]
```diff
@@ -1,101 +1,106 @@
...
 
-# choose the right platform init class
-def platform_sysv_init_class
-  value_for_platform_family(
-    'debian' => Chef::Provider::Service::Init::Debian,
-    'default' => Chef::Provider::Service::Init::Redhat
-  )
-end
+    # choose the right platform init class
+    def platform_sysv_init_class
```
Comment (tas50) [inline libraries/helpers.rb:87]: you probably don't need this if the sysv support is gone

File: resources/memcached_instance.rb (hunk 4 above)
Comment (ramereth) [reply resources/memcached_instance.rb:23]: Is this done by setting ``provides``? If so I'm having trouble getting the tests to pass by doing that. It seems as though it's not touching the ``systemd_unit`` resource. Is there a better way around this?

Comment (tas50) [reply resources/memcached_instance.rb:23]: Here's an example of a backwards compat name
https://github.com/chef-cookbooks/cron/blob/master/resources/access.rb#L26

File: CHANGELOG.md [hunk 6]
```diff
@@ -9,17 +9,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
-- Systemd service will be used when systemd is available if not then sysv init is used
+- Switched to using systemd_unit resource
 - migrated to github actions
+- Update and improved ChefSpec tests
+- Move library helper functions into a Memcached::Helpers namespace
 
 ### Removed
 
-- Support / testing for Ubuntu 14.04
+- Support / testing for Ubuntu 14.04, CentOS 6 & Debian 8
```
Comment (tas50) [inline CHANGELOG.md:27]: Debian 8 should probably just be in the testing section. It should work with the systemd stuff even if it's untested

File: README.md [hunk 7]
```diff
@@ -15,7 +15,6 @@ Provides a custom resource for installing instances of memcached. Also ships wit
 - Debian / Ubuntu and derivatives
 - RHEL and derivatives
 - Fedora
-- openSUSE/SLES
```
Comment (tas50) [inline README.md:18]: You might want to mention that we're not testing suse anymore

File: libraries/helpers.rb [hunk 8]
```diff
@@ -1,101 +1,100 @@
...
+      service 'disable default memcached' do
+        service_name 'memcached'
+        action [:stop, :disable]
+        not_if { new_resource.instance_name == 'memcached' }
+      end
+    end
 
-def remove_default_memcached_configs
-  return unless new_resource.remove_default_config
+    def remove_default_memcached_configs(new_resource)
```
Comment (tas50) [inline libraries/helpers.rb:48]: If you include these helpers into the action_class they should get the new_resource automatically

**Pull request merged**